x/shield/keeper/proposal.go

```diff
@@ -17,3 +24,415 @@ func (k Keeper) ClaimUnlock(ctx sdk.Context, proposalID, poolID uint64, loss sdk
+			purchased = totalPurchased
+		}
+		payout := provider.Collateral.ToDec().Mul(payoutRatio).TruncateInt()
+		if payout.GT(totalPayout) {
+			payout = totalPayout
+		}
+
+		// Require providers to cover (purchased + 1) and (payout + 1) if it's possible,
+		// so that the last provider will not be asked to cover all truncated amount.
+		if purchased.LT(totalPurchased) && provider.Collateral.GT(payout.Add(purchased)) {
+			purchased = purchased.Add(sdk.OneInt())
+		}
+		if payout.LT(totalPayout) && provider.Collateral.GT(payout.Add(purchased)) {
+			payout = payout.Add(sdk.OneInt())
+		}
+
+		if err := k.UpdateProviderCollateralForPayout(ctx, provider.Address, purchased, payout); err != nil {
+			panic(err)
+		}
+
+		if err := k.MakePayoutByProviderDelegations(ctx, provider.Address, purchased, payout); err != nil {
+			panic(err)
+		}
+
+		totalPurchased = totalPurchased.Sub(purchased)
+		totalPayout = totalPayout.Sub(payout)
+	}
+	reimbursement := types.NewReimbursement(rmb, beneficiary, ctx.BlockTime().Add(k.GetClaimProposalParams(ctx).PayoutPeriod))
+	k.SetReimbursement(ctx, proposalID, reimbursement)
+	return nil
+}
+
+// UpdateProviderCollateralForPayout updates a provider's collateral and withdraws according to the payout.
+func (k Keeper) UpdateProviderCollateralForPayout(ctx sdk.Context, providerAddr sdk.AccAddress, purchased, payout sdk.Int) error {
+	provider, found := k.GetProvider(ctx, providerAddr)
+	if !found {
+		return types.ErrProviderNotFound
+	}
+
+	uncoveredPurchase := sdk.ZeroInt()
+	payoutFromCollateral := sdk.ZeroInt()
+	if provider.Collateral.Sub(provider.Withdrawing).GTE(purchased.Add(payout)) {
+		// If collateral - withdraw >= purchased + payout:
+		//     purchased       payout
+		//   ----------------|--------|
+		//       collateral - withdraw                withdraw
+		// -------------------------------|---------------------------------
+		payoutFromCollateral = payout
+	} else if provider.Collateral.Sub(provider.Withdrawing).GTE(purchased) {
+		// If purchased <= collateral - withdraw < purchased + payout:
+		//               purchased       payout
+		//             ----------------|--------|
+		//       collateral - withdraw                withdraw
+		// -------------------------------|---------------------------------
+		payoutFromCollateral = provider.Collateral.Sub(provider.Withdrawing).Sub(purchased)
+	} else {
+		// If collateral - withdraw < purchased:
+		//                      purchased       payout
+		//                    ----------------|--------|
+		//       collateral - withdraw                withdraw
+		// -------------------------------|---------------------------------
+		uncoveredPurchase = purchased.Sub(provider.Collateral.Sub(provider.Withdrawing))
+	}
+	payoutFromWithdraw := payout.Sub(payoutFromCollateral)
+
+	// Update provider's collateral and total withdraw.
+	provider.Collateral = provider.Collateral.Sub(payout)
+	provider.Withdrawing = provider.Withdrawing.Sub(payoutFromWithdraw)
+
```

Comment:
We could early return here if `payoutFromWithdraw` is not positive.

Reply:
The check will be called once the for loop starts, so I think it's acceptable?